### PR TITLE
[NFC] Explain the order for adding module loaders.

### PR DESCRIPTION
**Pull request by varungandhi-apple**

### CODE REVIEW

**Review by harlanhaskins — APPROVED** [review]

File: lib/Frontend/Frontend.cpp [hunk 1]
```diff
@@ -353,10 +370,6 @@ bool CompilerInstance::setUpModuleLoaders() {
     Context->addModuleLoader(std::move(MemoryBufferLoader));
   }
 
-  std::unique_ptr<SerializedModuleLoader> SML =
-    SerializedModuleLoader::create(*Context, getDependencyTracker(), MLM);
-  this->SML = SML.get();
```
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: Huh, I wonder if this order mattered at one point.

File: lib/Frontend/Frontend.cpp [hunk 2]
```diff
@@ -317,6 +317,23 @@ void CompilerInstance::setUpDiagnosticOptions() {
...
+//    Ideally, we'd get rid of it.
+// 2. MemoryBufferSerializedModuleLoader: This is used by LLDB, because it might
+//    already have the module available in memory.
+// 3. ModuleInterfaceLoader: Tries to find an up-to-date swiftmodule. If it
+//    succeeds, it issues a particular "error" (see
+//    [Note: ModuleInterfaceLoader-defer-to-SerializedModuleLoader]), which
+//    is interpreted by the overarching loader as a command to use the
+//    SerializedModuleLoader. If we failed to find a .swiftmodule, this falls
+//    back to using an interface. Actual errors lead to diagnostics.
+// 4. SerializedModuleLoader: Tries to load the serialized module.
```
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: Nitpick: the "the" here is a little funny. What makes it "the" serialized module?

File: lib/Frontend/Frontend.cpp [hunk 3]
```diff
@@ -317,6 +317,23 @@ void CompilerInstance::setUpDiagnosticOptions() {
...
+// 5. ClangImporter: This must come after all the Swift module loaders because
+//    in the presence of overlays, we want to prefer the overlay over the
```
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: and mixed-source frameworks

**Review by jrose-apple — COMMENTED** [review]
Thanks for writing this out!

File: lib/Frontend/Frontend.cpp (hunk 2 above)
Comment (varungandhi-apple) [reply lib/Frontend/Frontend.cpp]: In `ASTContext::getModule`, we load one module at a time using `loadModule`, so I figured it should be "the" serialized module. How would you phrase it -- "a serialized module"?

File: lib/Frontend/Frontend.cpp (hunk 1 above)
Comment (varungandhi-apple) [reply lib/Frontend/Frontend.cpp]: Even if it did, I don't see a reason to keep it because it doesn't seem to matter now :).

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test and merge

File: lib/Frontend/Frontend.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/Frontend/Frontend.cpp]: That would work. The phrasing "the serialized module" assumes that every module we load has a serialized module, which isn't true. (The previous item also sorta falls prey to this.)

Comment (varungandhi-apple) [reply lib/Frontend/Frontend.cpp]: I've changed it to "Loads a serialized module if it can".

**Pull request merged**